File: CivitasJ/src/civitas/Jugador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
/**
 *
 * @author ramonjtorres
 */
public class Jugador implements Comparable<Jugador>{
    
    protected static int CasasMax = 4;
    protected static int CasasPorHotel = 4;
    protected boolean encarcelado;
    protected static int HotelesMax = 4;
    private String nombre;
    private int numCasillaActual;
    protected static float PasoPorSalida = 1000;
    protected static float PrecioLibertad = 200;
    private boolean puedeComprar;
    private float saldo;
    private static float SaldoInicial = 7500;
    ArrayList<TituloPropiedad> propiedades;
    Sorpresa salvoconducto;
    
    Jugador(String nombre){
        this.nombre = nombre;
        this.saldo = SaldoInicial;
        this.numCasillaActual = 0;
        this.propiedades = new ArrayList();
        this.puedeComprar = true;
        this.encarcelado = false;
        this.salvoconducto = null;
    }
    
    protected Jugador(Jugador otro){
        this.nombre = otro.nombre;
        this.numCasillaActual = otro.numCasillaActual;
        this.encarcelado = otro.encarcelado;
        this.propiedades = otro.propiedades;
        this.puedeComprar = otro.puedeComprar;
        this.salvoconducto = otro.salvoconducto;
        this.saldo = otro.saldo;
    }
    
    boolean cancelarHipoteca(int ip){return false;}
    
    int cantidadCasasHoteles(){
        
        int cantidad = 0;
        
        for(TituloPropiedad p:propiedades){
        
            cantidad += p.cantidadCasasHoteles();
        }
        
        return cantidad;
    }
    
    
    public int compareTo(Jugador otro){
    
        return Float.compare(this.saldo, otro.saldo);
    }
    
    boolean comprar(TituloPropiedad titulo){return false;}
    
    boolean construirCasa(int ip){return false;}
    
    boolean construirHotel(int ip){return false;}
    
    protected boolean debeSerEncarcelado(){
    
        if(this.encarcelado){
     
            return false;
        }
        else if(!this.tieneSalvoconducto()){
            return true;
        }
        else{
            this.perderSalvoConducto();
            Diario.getInstance().ocurreEvento("El jugador se libra de la carcel");
            return false;
        }
        
    }
    
    boolean enBancarrota(){
        return this.saldo < 0;
    }
    
    boolean encarcelar(int numCasillaCarcel){
        if(this.debeSerEncarcelado()){
            this.moverACasilla(numCasillaCarcel);
            this.encarcelado = true;
            Diario.getInstance().ocurreEvento("El jugador ha sido encarcelado");
        }
        return this.encarcelado;
    }
    
    private boolean existeLaPropiedad(int ip){
        
        return ip < this.propiedades.size();
    }
    
    private int getCasasMax(){
        return CasasMax;
    }
    
    int getCasasPorHotel(){
        return CasasPorHotel;
    }
    
    private int getHotelesMax(){
        return HotelesMax;
    }
    
    protected String getNombre(){
        return nombre;
    }
    
    int getNumCasillaActual(){
        return numCasillaActual;
    }
    
    private float getPrecioLibertad(){
        return PrecioLibertad;
    }
    
    private float getPremioPasoSalida(){
        return PasoPorSalida;
    }
    
    protected ArrayList<TituloPropiedad> getPropiedades(){
        return propiedades;
    }
    
    boolean getPuedeComprar(){
        return puedeComprar;
    }
    
    protected float getSaldo(){
        return saldo;
    }
    
    boolean hipotecar(int ip){return false;}
    
    public boolean isEncarcelado(){
        return encarcelado;
    }
    
    boolean modificarSaldo(float cantidad){
        this.saldo = this.saldo + cantidad;
        Diario.getInstance().ocurreEvento("Se ha modificado el saldo ahora tienes " + this.saldo);
        return true;
    }
    
    boolean moverACasilla(int numCasilla){
        if(this.encarcelado)
            return false;
        else{
            this.numCasillaActual = numCasilla;
            this.puedeComprar = false;
            Diario.getInstance().ocurreEvento("El jugador ha sido movido de casilla a la numero " + this.numCasillaActual);
            return true;
        }
    }
    
    boolean obtenerSalvoconducto(Sorpresa sorpresa){
        if(this.encarcelado)
            return false;
        this.salvoconducto = sorpresa;
        return true;
    }
    
    boolean paga(float cantidad){
        return this.modificarSaldo(-1*cantidad);
    }
    
    boolean pagaAlquiler(float cantidad){
        if(this.encarcelado)
            return false;
        else
            return this.paga(cantidad);
    }
    
    boolean pagaImpuesto(float cantidad){
        if(this.encarcelado)
            return false;
        else
            return this.paga(cantidad);
    }
    
    boolean pasaPorSalida(){
        this.modificarSaldo(1000);
        Diario.getInstance().ocurreEvento("El jugador ha pasado por salida");
        return true;
    }
    
    private void perderSalvoConducto(){
        this.salvoconducto.usada();
        this.salvoconducto = null;
    }
    
    boolean puedeComprarCasilla(){
        if(this.encarcelado)
            this.puedeComprar = false;
        else
            this.puedeComprar = true;
        return puedeComprar;
    }
    
    private boolean puedeSalirCarcelPagando(){
        return (this.saldo>this.getPrecioLibertad());
    }
    
    private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        
        if(this.propiedades.contains(propiedad) && this.saldo >= propiedad.getPrecioEdificar() && propiedad.getNumCasas() < 4){
        
            return true;
        }
        else{
        
            return false;
        }
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        
        if(this.propiedades.contains(propiedad) && this.saldo >= propiedad.getPrecioEdificar() && propiedad.getNumCasas() == 4 && propiedad.getNumHoteles() < 4){
        
            return true;
        }
        else{
        
            return false;
        }
    }
    
    private boolean puedoGastar(float precio){
        if(this.encarcelado)
            return false;
        else
            return(this.saldo>precio);
    }
    
    boolean recibe(float cantidad){
        if(this.encarcelado)
            return false;
        else
            return this.modificarSaldo(cantidad);
    }
    
    boolean salirCarcelPagando(){
        if(this.encarcelado){
         
            this.paga(this.getPrecioLibertad());
            this.encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador ha salido de la carcel pagando");
            return true;   
        }
        else{
        
            return false;
        }
    }
    
    boolean salirCarcelTirando(){
        if(Dado.getInstance().salgoDeLaCarcel()){
            this.encarcelado = false;
            Diario.getInstance().ocurreEvento("El jugador ha salido de la carcel tirando el dado");
            return true;
        }
        else return false;
    }
    
    boolean tieneAlgoQueGestionar(){
        
        if(this.propiedades.isEmpty()){
         
            return false;
        }
        else{
            
            return true;
        }
    }
    
    boolean tieneSalvoconducto(){
        return (this.salvoconducto != null);
    }
    
    boolean vender(int ip){
        
        if(this.encarcelado){
            
            return false;
        }
        else{
            
            if(this.existeLaPropiedad(ip)){
                
                if(propiedades.get(ip).vender(this)){
                 
                    Diario.getInstance().ocurreEvento("Se ha vendido la propiedad: " + propiedades.get(ip).toString());
                    propiedades.remove(ip);
                    return true;
                }
            }
            
            return false;
        }
    }
    
    public String toString(){
     
        String s = "No";
        
        if(this.tieneSalvoconducto()){
        
            s = "Sí";
        }
        
        return "Nombre: " + this.getNombre() +
               "\nSaldo: " + this.getSaldo() +
               "\nCasilla Actual: " + this.getNumCasillaActual() +
               "\nSalvoconducto: " + s +
               "\nEncarcelado: " + this.isEncarcelado() +
               "\nPuede comprar: " + this.getPuedeComprar() +
               "\nPropiedades: " + "\n" +
                propiedades.toString(); 

    }
    
    /**
     * @param args the command line arguments
     */
    public static void main(String[] args) {

        Jugador jugador = new Jugador("Ramón");
        Jugador jugador2 = new Jugador("David");
        Jugador jugador3 = new Jugador(jugador);
        
        System.out.println(jugador3.toString());
        
        System.out.println("No tiene algo que gestionar. Debe dar false: " + jugador.tieneAlgoQueGestionar());
        
        TituloPropiedad propiedad = new TituloPropiedad("Ronda de Valencia",10, (float) 0.5,25,50,20);
        
        jugador.propiedades.add(propiedad);
        
        propiedad.comprar(jugador);
        
        System.out.println("Precio compra: " + propiedad.getPrecioCompra());
        
        System.out.println("Precio edificación: " + propiedad.getPrecioEdificar());
        
        System.out.println("Tiene algo que gestionar. Debe dar true: " + jugador.tieneAlgoQueGestionar());
        
        System.out.println("Debe dar 0, no tiene casas ni hoteles: " + jugador.cantidadCasasHoteles());
        
        System.out.println("Debe dar true, puede edificar casas: " + jugador.puedoEdificarCasa(propiedad));
        
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        
        System.out.println("Debe dar false, no puede edificar casas: " + jugador.puedoEdificarCasa(propiedad));
        
        System.out.println("Debe dar 4, tiene 4 casas: " + jugador.cantidadCasasHoteles());
        
        System.out.println("Debe dar true, puede edificar hoteles: " + jugador.puedoEdificarHotel(propiedad));
        
        System.out.println("HOTEL CONSTRUIDO: " + jugador.propiedades.get(0).construirHotel(jugador));
        
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        
        System.out.println("Debe dar 2, tiene 1 hotel y 1 casa: " + jugador.cantidadCasasHoteles());
        
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        
        System.out.println("HOTEL CONSTRUIDO: " + jugador.propiedades.get(0).construirHotel(jugador));
        
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        
        System.out.println("HOTEL CONSTRUIDO: " + jugador.propiedades.get(0).construirHotel(jugador));
        
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        System.out.println("CASA CONSTRUIDA: " + jugador.propiedades.get(0).construirCasa(jugador));
        
        System.out.println("HOTEL CONSTRUIDO: " + jugador.propiedades.get(0).construirHotel(jugador));
        
        System.out.println("Debe dar false, no puede edificar hoteles: " + jugador.puedoEdificarHotel(propiedad));
        
        System.out.println("Debe dar 7050: " + jugador.getSaldo());
        
        jugador.modificarSaldo(-200);
        
        System.out.println("Debe dar 6850: " + jugador.getSaldo());
        
        System.out.println("Debe dar -1 ya que jugador tiene menos saldo: " + jugador.compareTo(jugador2));
        System.out.println("Debe dar 1 ya que jugador2 tiene mas saldo: " + jugador2.compareTo(jugador));
        
        System.out.println("Debe dar false, no tiene salvoconducto: " + jugador.tieneSalvoconducto());
        
        System.out.println("Debe dar true, debe ser encarcelado: " + jugador.debeSerEncarcelado());
        
        jugador.encarcelado = true;
        
        System.out.println("Debe dar true, esta encarcelado: " + jugador.isEncarcelado());
        
        System.out.println("Debe dar false, no puede comprar si esta encarcelado: " + jugador.puedeComprarCasilla());
        
        jugador.encarcelado = false;
        
        System.out.println("Debe dar true, no esta encarcelado, puede comprar: " + jugador.puedeComprarCasilla());
        
        System.out.println("Debe dar false, no esta encarcelado: " + jugador.isEncarcelado());
        
        MazoSorpresas mazo = new MazoSorpresas();
        
        jugador.obtenerSalvoconducto(new Sorpresa(TipoSorpresa.SALIRCARCEL , mazo));
        
        System.out.println("Debe dar false, no esta en bancarrota: " + jugador.enBancarrota());
        
        System.out.println("Debe dar true, tiene salvoconducto: " + jugador.tieneSalvoconducto());
        
        System.out.println("Debe dar false, no debe ser encarcelado: " + jugador.debeSerEncarcelado());
        
        jugador.encarcelar(4);
        
        System.out.println("Debe dar true, existe la propiedad: " + jugador.existeLaPropiedad(0));
        System.out.println("Debe dar false, no existe la propiedad: " + jugador.existeLaPropiedad(5));
        
        System.out.println("Debe dar 4 casas max: " + jugador.getCasasMax());
        System.out.println("Debe dar 4 casas por hotel: " + jugador.getCasasPorHotel());
        System.out.println("Debe dar 4 hoteles max: " + jugador.getHotelesMax());
        
        System.out.println("Debe dar 4, estamos en la carcel: " + jugador.getNumCasillaActual());
        
        System.out.println("Debe dar 200 de precio de libertad: " + jugador.getPrecioLibertad());
        
        System.out.println("Debe dar 1000 de premio de salida: " + jugador.getPremioPasoSalida());
        
        System.out.println("Debe dar los datos de Ronda de Valencia: " + jugador.getPropiedades().toString());
        
        System.out.println("Debe dar false, no puede comprar: " + jugador.getPuedeComprar());
        
        jugador.puedeComprar = true;
        
        System.out.println("Debe dar true, puede comprar: " + jugador.getPuedeComprar());
        
        jugador.encarcelado = false;
        jugador.numCasillaActual = 0;
        
        System.out.println("Debe dar true, se puede mover a la carcel: " + jugador.moverACasilla(4));
        
        jugador.encarcelado = true;
        
        System.out.println("Debe dar false, no se puede mover a la carcel: " + jugador.moverACasilla(4));
        
        System.out.println("Debe dar true, puede salir pagando: " + jugador.salirCarcelPagando());
        
        System.out.println("Debe dar false, no puede salir pagando si no esta en la carcel: " + jugador.salirCarcelPagando());
        
        System.out.println("Debe dar true, paga 200: " + jugador.paga(200));
        
        System.out.println("Debe dar true, paga 200 de alquiler: " + jugador.pagaAlquiler(200));
        
        System.out.println("Debe dar true, paga 200 de impuesto: " + jugador.pagaImpuesto(200));
        
        System.out.println("Debe dar 6050 de saldo: " + jugador.getSaldo());
        
        System.out.println("Debe dar true al pasar por la salida: " + jugador.pasaPorSalida());
        
        System.out.println("Debe dar 7050 de saldo: " + jugador.getSaldo());
        
        jugador.obtenerSalvoconducto(new Sorpresa(TipoSorpresa.SALIRCARCEL , mazo));
        
        System.out.println("Debe dar true, tiene salvoconducto: " + jugador.tieneSalvoconducto());
        
        jugador.perderSalvoConducto();
        
        System.out.println("Debe dar false, no tiene salvoconducto: " + jugador.tieneSalvoconducto());
        
        System.out.println("Debe dar true, puede salir pagando: " + jugador.puedeSalirCarcelPagando());
        
        System.out.println("Debe dar true, puede gastar 7000: " + jugador.puedoGastar(7000));
        
        System.out.println("Debe dar true, pierde 7000: " + jugador.modificarSaldo(-7000));
        
        System.out.println("Debe dar false, no puede salir pagando: " + jugador.puedeSalirCarcelPagando());
        
        System.out.println("Debe dar false, no puede gastar 7000: " + jugador.puedoGastar(7000));
        
        System.out.println("Debe dar 50 de saldo: " + jugador.getSaldo());
        
        System.out.println("Debe dar true al recibir 500: " + jugador.recibe(500));
        
        System.out.println("Debe dar 550 de saldo: " + jugador.getSaldo());
        
        System.out.println("Debe dar true en algun momento: " + jugador.salirCarcelTirando());
        System.out.println("Debe dar true en algun momento: " + jugador.salirCarcelTirando());
        System.out.println("Debe dar true en algun momento: " + jugador.salirCarcelTirando());
        System.out.println("Debe dar true en algun momento: " + jugador.salirCarcelTirando());
        System.out.println("Debe dar true en algun momento: " + jugador.salirCarcelTirando());
        System.out.println("Debe dar true en algun momento: " + jugador.salirCarcelTirando());
        
        jugador.encarcelado = false;
        
        System.out.println("Debe dar true al vender: " + jugador.vender(0));
        
        System.out.println("Debe dar null: " + jugador.getPropiedades().toString());
        
        
        System.out.println("Debe dar false al vender: " + jugador.vender(0));
        
        System.out.println("Debe dar los datos del jugador: " + jugador.toString());
        
        System.out.println("Debe dar true al modificar -10000 de saldo: " + jugador.modificarSaldo(-10000));
        
        System.out.println("Debe dar true, esta en bancarrota: " + jugador.enBancarrota());
        
    }
}

File: CivitasR/lib/jugador.rb
#encoding: utf-8

# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Jugador
    include Comparable
    
    attr_reader :casas_max, :hoteles_max, :casas_por_hotel, :nombre, :num_casilla_actual, :precio_libertad, :paso_por_salida, :propiedades, :puede_comprar, :saldo
    
    @@casas_max = 4
    @@hoteles_max = 4
    @@casas_por_hotel = 4
    @@paso_por_salida = 1000
    @@precio_libertad = 200
    @@saldo_inicial = 7500
    
    public
    def initialize(nombre)
    
      @encarcelado = false
      @num_casilla_actual = 0
      @puede_comprar = true
      @saldo = @@saldo_inicial
      @nombre = nombre
      @propiedades = Array.new
      @salvoconducto = nil
      
    end
    
    protected
    def jugador(otro)
      
      @nombre = otro.nombre
      @num_casilla_actual = otro.numCasillaActual
      @encarcelado = otro.encarcelado
      @propiedades = otro.propiedades
      @puede_pomprar = otro.puedeComprar
      @salvoconducto = otro.salvoconducto
      @saldo = otro.saldo
      
    end
    
    public
    def cancelar_hipoteca(ip)
      
      result = false
        
      if(@encarcelado)

          return result
      end

      if(existeLaPropiedad(ip))

          propiedad = @propiedades.at(ip)
          cantidad = propiedad.get_importe_cancelar_hipoteca()
          puedogastar = puedo_gastar(cantidad)

          if(puedogastar)

              result = propiedad.cancelar_hipoteca(self)

              if(result)

                  Diario.instance.ocurre_evento("El jugador" + @nombre + "cancela la hipoteca de la propiedad" + ip)
              end
          end
      end

      return result
      
    end
    
    public
    def cantidad_casas_hoteles()
      
      cantidad = 0
      
      @propiedades.each do |p|
        
            cantidad += p.cantidad_casas_hoteles();
      end
      
      return cantidad
      
    end
    
    public
    def compare_to(otro)
      
      return (@saldo <=> otro.saldo)
    end
    
    public
    def comprar(titulo)
      
      result = false

      if(@encarcelado)

          return result
      end

      if(@puede_comprar)

          precio = titulo.precio_compra

          if(puedo_gastar(precio))

              result = titulo.comprar(self)

              if(result)

                  @propiedades.push(titulo)
                  Diario.instance.ocurre_evento("El jugador" + self + "compra la propiedad" + titulo.to_s)
              end
          end
      end

      return result
    end
    
    public
    def construir_casa(ip)
      
    end
    
    public
    def construir_hotel(ip)
      
    end
    
    protected
    def debe_ser_encarcelado()
      if(@encarcelado)
        return false
      elsif(!tiene_salvoconducto())
        return true
      else
        perder_salvoconducto()
        Diario.instance.ocurre_event("El jugador se libra de la carcel")
        return false
      end
      
    end
    
    public
    def en_bancarrota()
      
      return @saldo < 0      
    end
    
    public
    def encarcelar(num_casilla_carcel)
      if(debe_ser_encarcelado())
        mover_a_casilla(num_casilla_carcel)
        @encarcelado = true
        
        Diario.instance.ocurre_evento("El jugador ha sido encarcelado")
      end
      return @encarcelado
      
    end
    
    private
    def existe_la_propiedad(ip)
      
      return ip < @propiedades.length()
      
    end
    
    public
    def hipotecar(ip)
      
      result = false

      if(@encarcelado)

          return result
      end

      if(existe_la_propiedad(ip))

          propiedad = @propiedades.at(ip)
          result = propiedad.hipotecar(self)

          if(result)

              Diario.instance.ocurre_evento("El jugador" + @nombre + "hipoteca la propiedad" + ip)
          end
      end

      return result
    end
    
    public
    def is_encarcelado()
      
      return @encarcelado
    end
    
    public
    def modificar_saldo(cantidad)
      
      diario = Diario.instance
      
      @saldo = @saldo + cantidad
      diario.ocurre_evento("Se ha modificado el saldo ahora tienes "+@saldo.to_s)
      return true
      
    end
    
    public
    def mover_a_casilla(num_casilla)
      if(@encarcelado)
        return false
      else
        diario = Diario.instance
        
        @num_casilla_actual = num_casilla
        @puede_comprar = false
        diario.ocurre_evento("El jugador a sido movido de casilla a la numero "+ @num_casilla_actual.to_s)
        return true
      end
      
    end
    
    public
    def obtener_salvoconducto(sorpresa)
      if(@encarcelado)
        return false
      end
      @salvoconducto = sorpresa
      return true
    end
    
    public
    def paga(cantidad)
      modificar_saldo(-1*cantidad)
    end
    
    public
    def paga_alquiler(cantidad)
      if(@encarcelado)
        return false
      else
        return paga(cantidad)
      end
    end
    
    public
    def paga_impuesto(cantidad)
      if(@encarcelado)
        return false
      else
        return paga(cantidad)
      end
    end
    
    public
    def pasa_por_salida()
      modificar_saldo(1000)
      Diario.instance.ocurre_evento("El jugador ha pasado por salida")
      return true
      
    end
    
    private
    def perder_salvoconducto()
      
      @salvoconducto.usada()
      @salvoconducto = nil
      
    end
    
    public
    def puede_comprar_casilla()
      if(@encarcelado)
        @puede_comprar = false
      else
        @puede_comprar = true
      end
      return @puede_comprar
      
    end
    
    private
    def puede_salir_carcel_pagando()
      return @saldo>get_precio_libertad()
      
    end
    
    private
    def puedo_edificar_casa(propiedad)
      
      if(@propiedades.include?(propiedad) && @saldo >= propiedad.precio_edificar && propiedad.num_casas < 4)
       
        return true
        
      else
        
        return false
        
      end
      
    end
    
    private
    def puedo_edificar_hotel(propiedad)
      
      if(@propiedades.include?(propiedad) && @saldo >= propiedad.precio_edificar && propiedad.num_casas == 4 && propiedad.num_hoteles < 4)
       
        return true
        
      else
        
        return false
        
      end
      
    end
    
    private
    def puedo_gastar(precio)
        if(@encarcelado)
          return false
        else
          return (@saldo>precio)
        end
      
    end
    
    public
    def recibe(cantidad)
      if(@encarcelado)
        return false
      else
        return modificar_saldo(cantidad)
      end
      
    end
    
    public
    def salir_carcel_pagando()
      if(@encarcelado)
        paga(@@precio_libertad)
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugado ha salido de la carcel pagando")
      end
      
    end
    
    public
    def salir_carcel_tirando()
      if(Dado.instance.salgo_de_la_carcel())
        @encarcelado = false
        Diario.instance.ocurre_evento("El jugador ha salido de la carcel tirando el dado")
        return true
      end
    end
    
    public
    def tiene_algo_que_gestionar()
      if(@propiedades.empty?)
        return false
      end
      return true
      
    end
    
    public
    def tiene_salvoconducto()
      return (@salvoconducto != nil)
    end
    
    public
    def vender(ip)
      if(@encarcelado)
        return false
      else
        if(existe_la_propiedad(ip))
          if(@propiedades.at(ip).vender(self))
            @propiedades.delete(ip)
            Diario.instance.ocurre_evento("Se ha vendido la propiedad: "+@propiedades.at(ip).to_s)
            return true
          end
        end
      end
      
    end
    
    public
    def to_s()
      
      s = "No"
        
      if(tiene_salvoconducto())
        
        s = "Sí"
      end
        
      return "Nombre: " + @nombre.to_s +
             "\nSaldo: " + @saldo.to_s +
             "\nCasilla Actual: " + @num_casilla_actual.to_s +
             "\nSalvoconducto: " + s +
             "\nEncarcelado: " + @encarcelado.to_s +
             "\nPuede comprar: " + @puede_comprar.to_s +
             "\nPropiedades: " + "\n" +
             @propiedades.to_s
      
    end
 
  end
  
end
